Askrene: prune and cap

Lagrang3 opened the pull request:
This PR depends on #8299 and supersedes #8314.

With this PR I want to make a small improvement to the MCF solver in askrene.

First: I would like to constrain the number of flow units to 1000 by setting the accuracy of the solver
to the total payment amount divided by 1000. Some MCF algorithms like "successive shortest path" (SSP)
have theoretical complexity bounds that depend on that number.
Note: the 1000 number is arbitrary, the smaller it is we may reduce the solver's runtime but we lose
a accuracy.

Second: I would like to prune the set of arcs in the network. I can achieve this by setting a limit to the sum
of the arc capacities that correspond to the same channel to U, the maximum number of flow units
in the payment. Notice that due to the piece-wise linearization of the channel
cost function, one channel becomes several arcs in the MCF network, therefore we can discard the higher cost
arcs of a channel linearization if the lower cost arcs already sum up to U in flow capacity.

Lagrang3 commented:
Runtime distribution for precision=1
![runtime-big](https://github.com/user-attachments/assets/3f156bd8-1432-4e69-b3a7-5c6e2ab17187)

Runtime distribution for precision=10
![runtime-big](https://github.com/user-attachments/assets/30c5909a-71e8-46e8-91de-1d267df3414c)

Runtime distribution for precision=100
![runtime-big](https://github.com/user-attachments/assets/652eeeb8-a23c-438d-9204-5b8047aa5736)

Runtime distribution for precision=1000
![runtime-big](https://github.com/user-attachments/assets/d055f650-c20b-4de1-b90c-9b0e619567b2)

Runtime distribution for precision=1M
![runtime-big](https://github.com/user-attachments/assets/0aa79949-5f6e-4d8f-9744-ee77738725a5)

For 1M the improvement in runtime are exclusively due to the pruning.

Lagrang3 commented:
![overtime](https://github.com/user-attachments/assets/afb8b835-7760-43c3-b4d0-8decbd9c6486)

rustyrussell reviewed (approved):
Minor nit:

Can you split 1b5e87a8fd9df1b9034b1c6f10f1182836c98240 into the "prune small channels" change, and the "dynamic granularity based on < 1000 flows assumption" (which, FWIW, is a *LOT* of flows!).

They're logically separate improvements...

rustyrussell commented:
Hmm, so this is a lot of data.  A summary of this in the commit message itself would be invaluable:   Remember, GitHub is transitory, and will one day stop hosting us for free: git commits are forever.

You graph precision amounts, but then if I understand correctly, your precision varies on the total payment amount, so these graphs don't exactly apply to our overall performance?

Lagrang3 commented:
We also need a better commit message explaining what we are actually pruning.
Not that we are pruning small channels but arcs that we can prove are never going to be needed
in the optimal MCF solution, why? because there are other arcs with lower cost between the same
pair of nodes that combined have a capacity greater or equal the total flow amount.

Lagrang3 commented:
- split into three commits,
- added some benchmarks results to the commit description,
- added comment in the code better explaining what is pruned,
- rebased into #8299,
- and the pytests instances for every commit.

rustyrussell commented:
Trivial rebase now predecessor merged!
